include/zephyr/net/net_if.h

```diff
@@ -1366,13 +1366,18 @@ void net_if_ipv6_addr_foreach(struct net_if *iface, net_if_ip_addr_cb_t cb,
 /**
  * @brief Add a IPv6 multicast address to an interface
  *
+ * Note that this function runs a net_if_ipv6_maddr_lookup() internally.
+ *
  * @param iface Network interface
  * @param addr IPv6 multicast address
+ * @param ifmaddr A pointer to a valid struct net_if_mcast_addr pointer holder
  *
- * @return Pointer to interface multicast address, NULL if cannot be added
+ * @return 0 on sucses, a negative errno otherwise. Note that -EALREADY also
```

Comment:
Maybe just align the behavior with the unicast `net_if_ipv6_addr_add()`? That function does the lookup internally, and just silently skips address registration if it already exists (and still returns `ifaddr` pointer)? That way no API changes would be needed.

Reply:
indeed, after all net_if_ipv6_maddr_add() loops twice on the multicast table... so it would also optimize this out.

Reply:
Unlike net_if_ipv6_addr_add(), current net_if_ipv6_maddr_add() return NULL in case of no memory or if it was already there. So, yes some inconsistency in the API. That explains why users have been calling net_if_ipv6_maddr_lookup() before: to be able to mitigate if maddr is already there etc... Thus also why some tests assumes EALREADY in case of NULL returned.

Note that the documentation of the function is ambiguous enough that it could let us change the behavior to mimic net_if_ipv6_addr_add() (which has exactly the same documentation on its return value).

It just means that tests will need to change accordingly